### PR TITLE
Support comparing Sparse Compressed tensors

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #77530
* #77512
* __->__ #77525
* #77508
* #77504

### CODE REVIEW

File: torch/testing/_comparison.py [hunk 1]
```diff
@@ -602,7 +602,9 @@ def _to_tensor(self, tensor_like: Any) -> torch.Tensor:
             raise UnsupportedInputs()
 
     def _check_supported(self, tensor: torch.Tensor, *, id: Tuple[Any, ...]) -> None:
-        if tensor.layout not in {torch.strided, torch.sparse_coo, torch.sparse_csr}:  # type: ignore[attr-defined]
+        supported_layouts = {torch.strided, torch.sparse_coo, torch.sparse_csr, torch.sparse_csc,
+                             torch.sparse_bsr, torch.sparse_bsc}
+        if tensor.layout not in supported_layouts:  # type: ignore[attr-defined]
```
Comment (pmeier) [inline torch/testing/_comparison.py:607]: Given that we don't use `supported_layouts` anywhere else, can we remove it?

```suggestion
        if tensor.layout not in {
            torch.strided,
            torch.sparse_coo,
            torch.sparse_csr,
            torch.sparse_csc,
            torch.sparse_bsr,
            torch.sparse_bsc,
        }:
```

I've omitted the `type: ignore` comment here. Please only re-add it if `mypy` complains.

File: torch/testing/_comparison.py [hunk 2]
```diff
@@ -771,42 +773,59 @@ def _compare_sparse_coo_values(
...
-        - the crow_indices for equality, and
+        - the plain indices for equality,
+        - the compressed indices for equality, and
         - the values for closeness.
         """
+        format_name = {
+            torch.sparse_csr: 'CSR',
+            torch.sparse_csc: 'CSC',
+            torch.sparse_bsr: 'BSR',
+            torch.sparse_bsc: 'BSC'}[actual.layout]
```
Comment (pmeier) [inline torch/testing/_comparison.py:790]: Can we maybe merge the three dictionaries into one, i.e.

```py
{layout: (format_name, compressed_indices_mth, plain_indices_mth)}
```

File: torch/testing/_comparison.py [hunk 3]
```diff
@@ -771,42 +773,59 @@ def _compare_sparse_coo_values(
...
+            torch.sparse_csr: torch.Tensor.col_indices,
+            torch.sparse_csc: torch.Tensor.row_indices,
+            torch.sparse_bsr: torch.Tensor.col_indices,
+            torch.sparse_bsc: torch.Tensor.row_indices}[actual.layout]
+
         self._compare_regular_values_equal(
-            actual.crow_indices(),
-            expected.crow_indices(),
-            identifier="Sparse CSR crow_indices",
+            compressed_indices_mth(actual),
```
Comment (pmeier) [inline torch/testing/_comparison.py:813]: Is `mth` a sparse term? Otherwise, could you explain or replace with something more expressive?

**Review by pmeier — COMMENTED** [review]
Changes look good, thanks @pearu. In addition to my comments below, could you also adapt the docstring?

https://github.com/pytorch/pytorch/blob/6fa20bdfe85fe6b92744168edb1495c8e7545f9a/torch/testing/_comparison.py#L1109-L1111

Plus, some tests for the new layouts would also help ensure the behavior is correct:

https://github.com/pytorch/pytorch/blob/6fa20bdfe85fe6b92744168edb1495c8e7545f9a/test/test_testing.py#L995

https://github.com/pytorch/pytorch/blob/6fa20bdfe85fe6b92744168edb1495c8e7545f9a/test/test_testing.py#L1088

File: torch/testing/_comparison.py (hunk 2 above)
Comment (pearu) [reply torch/testing/_comparison.py:790]: sure

File: torch/testing/_comparison.py (hunk 3 above)
Comment (pearu) [reply torch/testing/_comparison.py:813]: `mth` is an abbreviation for `method`.

Comment (pmeier) [reply torch/testing/_comparison.py:813]: Ah :slightly_smiling_face: IMO the extra three characters will not hinder readability here and might help someone else reading this to not stumble over the same. No strong opinion though.

File: torch/testing/_comparison.py (hunk 1 above)
Comment (pearu) [reply torch/testing/_comparison.py]: Done.

File: torch/testing/_comparison.py (hunk 2 above)
Comment (pearu) [reply torch/testing/_comparison.py]: Done.

File: torch/testing/_comparison.py (hunk 3 above)
Comment (pearu) [reply torch/testing/_comparison.py]: Using `method`.

File: test/test_testing.py [hunk 4]
```diff
@@ -1145,6 +1145,189 @@ def test_mismatching_values_msg(self):
                 fn()
 
 
+@unittest.skipIf(IS_FBCODE or IS_SANDCASTLE, "Not all sandcastle jobs support CSC testing")
```
Comment (pmeier) [inline test/test_testing.py:1148]: Just confirming: is this true or did you just copy-paste it from the CSR test case?

**Review by pmeier — APPROVED** [review]
LGTM, thanks @pearu!

File: test/test_testing.py (hunk 4 above)
Comment (pearu) [reply test/test_testing.py:1148]: It was just a copy-paste. I would not expect that CSR and CSC tests would behave differently.

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**